pkg/crd/crd.go

```diff
@@ -18,58 +18,42 @@ package crd
 
 import (
 	"context"
-	"time"
+	"fmt"
 
+	"github.com/hashicorp/go-multierror"
 	"go.uber.org/zap"
-	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
 	apiextensionsclient "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
-	k8serrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-var functionCrdVersions, environmentCrdVersions, packageCrdVersions, httpTriggerCrdVersions, timeTriggerCrdVersions, k8swatchTriggerCrdVersions, mqTriggerCrdVersions, canaryconfigCrdVersions []apiextensionsv1.CustomResourceDefinitionVersion
-
-const (
-	crdGroupName = "fission.io"
-	crdVersion   = "v1"
-)
-
-// ensureCRD checks if the given CRD type exists, and creates it if
-// needed. (Note that this creates the CRD type; it doesn't create any
-// _instances_ of that type.)
-func ensureCRD(logger *zap.Logger, clientset *apiextensionsclient.Clientset, crd string) (err error) {
-	maxRetries := 5
-
-	for i := 0; i < maxRetries; i++ {
-		_, err = clientset.ApiextensionsV1().CustomResourceDefinitions().Get(context.TODO(), crd, metav1.GetOptions{})
-		if err == nil {
-			return nil
-		}
+// EnsureFissionCRDs checks if all Fission CRDs are present
+func EnsureFissionCRDs(logger *zap.Logger, clientset *apiextensionsclient.Clientset) error {
+	crds, err := clientset.ApiextensionsV1().CustomResourceDefinitions().List(context.TODO(), metav1.ListOptions{})
```

Comment:
Is it not simpler to just `Get()` each CRD as we have the names stored anyway in `crdsExpected`?  If the `Get()` call fails , we can append to `multierrror`

Reply:
LIst call fetches all CRDs in go compared to `Get` calls. So preferred it. Since we not apply CRDs via `kubectl create` Its expected user should have applied it, if not user just need to restart pod otherwise it would be in CrashLoop.

Reply:
Users could have a lot of CRDs in the cluster, in which case a `Get` call could be cheaper. Is it possible to use a field selector or label selector in the `List()` to narrow down the fission CRDs and just count if there are 8 of them?